### PR TITLE
Add support for Morpho vaults

### LINKED ISSUE #8618: Support Morpho Vaults
## Abstract

Support Morpho Vaults across all EVM chains

## Specification

Do it

### PR DESCRIPTION
**Pull request by nicholasyoder**
Closes #8618

### CODE REVIEW

**Review by yabirgb — COMMENTED** [review]
Question. This doesn´t add logic to query balances. Are the positions detected as erc tokens? The prices for what I saw are queried

File: rotkehlchen/chain/evm/decoding/morpho/constants.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from typing import Final
+
+from eth_typing import ABI
+
+CPT_MORPHO: Final = 'morpho'
+MORPHO_VAULT_ABI: ABI = [
+  {
+    'inputs': [
+      {
```
[the remark this answers is not in the document]
Comment (nicholasyoder) [reply rotkehlchen/chain/evm/decoding/morpho/constants.py]: Made it one line. Seems simplest just to have it here in constants, as is also done [here](https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/arbitrum_one/modules/umami/constants.py#L9) for example, but we could put the whole ABI in the globaldb if you think that would be better.

File: rotkehlchen/chain/evm/decoding/morpho/utils.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                timeout=CachedSettings().get_timeout_tuple(),
+            )
+            vault_list = response_data.json()['data']['vaults']['items']
+            all_vaults.extend(vault_list)
+            offset += VAULT_QUERY_PAGE_SIZE
+            if len(vault_list) < VAULT_QUERY_PAGE_SIZE:
+                break  # no more vaults to retrieve
+
+        except (requests.exceptions.RequestException, JSONDecodeError, TypeError, KeyError) as e:
+            error = f'missing key {e!s}' if isinstance(e, KeyError) else f'{e!s}'
```
[the remark this answers is not in the document]
Comment (nicholasyoder) [reply rotkehlchen/chain/evm/decoding/morpho/utils.py]: You're right about `requests.RequestException`, but when I tried testing with bad json it did in fact raise `json.decoder.JSONDecodeError`, so I left that the way it is.

**Comment by nicholasyoder** [comment]
> Question. This doesn´t add logic to query balances. Are the positions detected as erc tokens? The prices for what I saw are queried

Yes, the positions are detected as erc tokens. I was confused on this myself at first, and tried to add balances logic, which then made it show double the actual balance in the app when I tested it.

File: rotkehlchen/chain/ethereum/modules/morpho/decoder.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class MorphoDecoder(MorphoCommonDecoder):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/morpho/decoder.py:16]: nitpick: this should be ethnodeInquirer

File: rotkehlchen/chain/base/modules/morpho/decoder.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class MorphoDecoder(MorphoCommonDecoder):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
```
Comment (yabirgb) [inline rotkehlchen/chain/base/modules/morpho/decoder.py:17]: nitpic: this should be base node inquirer

File: rotkehlchen/tests/unit/test_inquirer.py [hunk 5]
```diff
@@ -934,6 +935,19 @@ def test_find_vthor_price(inquirer_defi: 'Inquirer', database: 'DBHandler'):
...
 
 
+@pytest.mark.vcr(filter_query_parameters=['apikey'])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('should_mock_current_price_queries', [False])
+def test_find_morpho_vault_price(
+        database: 'DBHandler',
+        inquirer_defi: 'Inquirer',
+) -> None:
+    """Test that we get the correct price for Morpho vault tokens."""
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_inquirer.py:945]: nitpick: doesn´t this fit in one line?

File: rotkehlchen/chain/evm/decoding/morpho/utils.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        return
+
+    _update_cache_timestamp(count=(vault_count := len(vault_list)))
+
+    try:
+        if last_vault_count is not None and vault_count == int(last_vault_count):
+            log.debug(f'Same number ({vault_count}) of Morpho vaults returned from API as previous query. Skipping vault processing.')  # noqa: E501
+            return
+    except ValueError as e:
+        log.error(f'Failed to check last Morpho vault count due to {e!s}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/morpho/utils.py:100]:
```suggestion
        log.error(f'Failed to check last Morpho vault count due to {last_vault_count} not being an int)
```

**Review by LefterisJP — COMMENTED** [review]
As mentioned in discord the missing things are:

- Prices for my vaults when testing do not appear
- The bundler transactions do not get decoded properly

File: rotkehlchen/chain/evm/decoding/morpho/utils.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                underlying_tokens=[UnderlyingToken(
+                    address=underlying_token.evm_address,
+                    token_kind=EvmTokenKind.ERC20,
+                    weight=ONE,
+                )],
+                encounter=TokenEncounterInfo(description='Querying Morpho vaults', should_notify=False),  # noqa: E501
+            )
+        except (DeserializationError, KeyError) as e:
+            error = f'missing key {e!s}' if isinstance(e, KeyError) else f'{e!s}'
+            log.error(f'Failed to store token information for Morpho vault due to {error}. Vault: {vault}. Skipping...')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/morpho/utils.py:132]: here you forgot to log which chain it is in. Add chain name

```suggestion
            log.error(f'Failed to store token information for Morpho vault due to {error}. Vault: {vault}. Skipping...')  # noqa: E501
```

File: rotkehlchen/chain/evm/decoding/morpho/utils.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        return ZERO_PRICE
+
+    if (
+        len(vault_token.underlying_tokens) == 0 or
+        (underlying_token := get_token(
+            evm_address=vault_token.underlying_tokens[0].address,
+            chain_id=evm_inquirer.chain_id,
+        )) is None
+    ):
+        log.error(f'Failed to get underlying token for Morpho vault {vault_token}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/morpho/utils.py:161]: same here for the chain

File: rotkehlchen/chain/evm/decoding/morpho/utils.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    try:
+        price_per_share = evm_inquirer.call_contract(
+            contract_address=vault_token.evm_address,
+            abi=MORPHO_VAULT_ABI,
+            method_name='convertToAssets',
+            arguments=[int(1e18)],  # 1 share with 18 decimals
+        )
+    except RemoteError as e:
+        log.error(f'Failed to get price per share for Morpho vault {vault_token}: {e}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/morpho/utils.py:151]: same here and in all other parts of this file

File: rotkehlchen/chain/evm/decoding/morpho/utils.py (hunk 7 above)
Comment (nicholasyoder) [reply rotkehlchen/chain/evm/decoding/morpho/utils.py:132]: This is processing morpho vaults from all chains. While we could use `vault['chain']['id']` to get the chain for the current vault here, deserializing that value is one of the things that could raise this error in the first place. 

Also the vault data is already being logged with `Vault: {vault}`, which will include the vault's chain id and name.

File: rotkehlchen/tests/unit/test_inquirer.py [hunk 10]
```diff
@@ -934,6 +935,16 @@ def test_find_vthor_price(inquirer_defi: 'Inquirer', database: 'DBHandler'):
     assert price.is_close(FVal(0.97656), max_diff=1e-5)
 
 
+@pytest.mark.vcr(filter_query_parameters=['apikey'])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('should_mock_current_price_queries', [False])
+def test_find_morpho_vault_price(database: 'DBHandler', inquirer_defi: 'Inquirer') -> None:
+    """Test that we get the correct price for Morpho vault tokens."""
+    usual_boosted_usdc_vault = create_ethereum_morpho_vault_token(database=database)
+    price = inquirer_defi.find_usd_price(asset=usual_boosted_usdc_vault)
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_inquirer.py:944]: So as mentioned in discord here if you do a EUR price query it won't work. But this is a deeper error which we have to handle.

**Review by LefterisJP — COMMENTED** [review]
Small comment and needs rebase and conflict resolution.

**Review by LefterisJP — APPROVED** [review]
LGTM

**Pull request merged**